tests/pyfunc/test_scoring_server.py

```diff
@@ -699,3 +734,83 @@ def test_scoring_server_client(sklearn_model, model_path):
     finally:
         if server_proc is not None:
             os.kill(server_proc.pid, signal.SIGTERM)
+
+
+@pytest.mark.parametrize("key", ["messages"])
+def test_scoring_server_allows_payloads_with_llm_keys_for_pyfunc(model_path, key):
```

Comment:
Awesome! Can we also include a signature in one of these test cases that matches OpenAI? This will help us ensure that the OpenAI-formatted payload can be used with a model that has a signature (required for Databricks UC Model Registry).

(I know you mentioned this testing offline, just leaving this comment here in case)

Reply:
@dbczumar @serena-ruan @daniellok-db 
Hmm two issues turned out when testing with signature, both are related to `_enforce_schema` function.

**Problems**
1. When a dictionary who has at least one Array or Object value is passed to `_enforce_schema`, it is wrapped by a list first so it can be converted to Pandas Dataframe as a single row. Imagine user creates a Pyfunc model that calls a proprietary API inside (with signature). The user need to add lines to convert it from pandas to dict like `if isinstance(model_input, pd.DataFrame): model_input = model_input.to_dict(orient="records")[0]`. The true painpoint here is that this error will never be caught until the user loads/serves it as pyfunc, and the error message will not be helpful as well.
2.  Current implementation of `_enforce_schema` complains when there is missing input, but [not for extra input](https://github.com/mlflow/mlflow/blob/master/mlflow/models/utils.py#L635), instead just discarding them. Hence although we adding those undefined params to input data in `_split_data_and_params_for_llm_input`, those params are thrown away here. To workaround this, users need to define input schema that includes params, rather than using params schema (defining them as params doesn't work because then those values need to be wrapped in a key "params", breaking unified interface).

**Short-term**
One option is to tweak schema enforcement logic similarly to what we have done - check if "message" is in the input and do special handling - but it will definitely come back to bite later. Good thing is that these two issues at least have workarounds. so I would just add warnings and documentation for now.
1. In [pyfunc.save_model()](https://github.com/mlflow/mlflow/blob/master/mlflow/pyfunc/__init__.py#L1729), check if the passed/inferred signature has "message" field with Array value, then show warning like `Warining: When an MLflow Python model saved with a signature makes prediction, it will first convert input data to Pandas Dataframe. You may need to unwrap it with `model_inpuit.to_dict(oriend="records")[0]` inside your custom predict function.
2. Inside `_split_data_and_params_for_llm_input`, when we find params not defined in schema, raise Exception and prompt users to define everything in input_data (when they want to make prediction with raw dict). It may by a bit annoying but definitely better than silently making wrong inference.

**Long-term**
That's said, none of above option is user-friendly - must be revised to long-term design, ideally in Q1. It needs more careful investigation but these are what I think doable:
1. I think we should make `_enforce_schema` to return the same type as its input, rather than DataFrame. Many flavors such as [openai](https://github.com/mlflow/mlflow/blob/master/mlflow/openai/__init__.py#L694), [langchain](https://github.com/mlflow/mlflow/blob/master/mlflow/langchain/__init__.py#L485), [transformer](https://github.com/mlflow/mlflow/blob/master/mlflow/tensorflow/__init__.py#L781), [pytorch](https://github.com/mlflow/mlflow/blob/master/mlflow/pytorch/__init__.py#L731) implements unwrapping logic due to this conversion. (more longer term... we may wanna remove pandas dependency for LLM case - 2D table is not suitable to LLM input that has more nested structure...)
2. (1) enforce_schema should not simply discard extra inputs - either include to data (bypass schema) or raise validation error (probably allow users to configure). (2) afaik params concept was introduced mainly for LLM, so if the unified interface is likely the dominant format of LLM inference, we may need to consider removing params before the logic becomes more complex.

What do you guys think about these? Long-term ones are way too early to discuss here and I think it's totally fine for now to say "the new unified interface doesn't work perfectly with signature" tho:p

Reply:
Thanks @B-Step62 ! Regarding the short term:

1. Since this behavior of converting to pandas is preexisting behavior for all models with ColSpec signatures, I don't think we need to print a warning. Users can add type hints to `inputs` if they want to receive a dictionary instead (see https://github.com/mlflow/mlflow/pull/10742#pullrequestreview-1796204890).

2. I don't think we should throw when an extra param is defined for the `messages` case. I think we should keep the behavior consistent with what we do for `dataframe_records` / `dataframe_split` / etc. For now, if users want to add extra params beyond what OpenAI supports, those extra params should be included in the signature.

Can we revert that commit? LGTM once commit is reverted

Reply:
Thanks @dbczumar, makes sense for the warning, will revert it. 
As for extra params, 
>  For now, if users want to add extra params beyond what OpenAI supports, those extra params should be included in the signature

what we have after LLM signiture doesn't align with this. Basically _enforce_schema drops extra keys not defined in the **input schema**, as it iterates over keys in input_schema to create pandas Dataframe ([code](https://github.com/mlflow/mlflow/blob/master/mlflow/models/utils.py#L635)). To avoid this, we need to modify _enforce_schema, which I'm a bit worrylng about breaking existing behavior.

Alternatively, we can have function like _enforce_cols_schema_llm or add flag like drop_extra, which is obviously not preferable but probably works best for users now.